14_hash_table/hash_table.js

```diff
@@ -0,0 +1,117 @@
+'use strict';
+
+const LinkedList = require('../04_linked_list/lib/linked_list');
+
+class HashTable {
+  constructor(nBuckets) {
+    if (!nBuckets) { throw new Error('No buckets number specified') }
+    this.nBuckets = nBuckets;
+    this.buckets = Array(nBuckets).fill(null);
+  }
+
+  multHash(key) {
+    if (!key) { throw new Error('no key provided for hashing') }
+
+    return key.split('').reduce((acc, curr) => {
+      return acc * curr.charCodeAt(0);
+    }, 1) % this.nBuckets;
+  }
+
+  set(key, value) {
+    if (arguments.length < 2) { throw new Error('set requires key and value') }
+    let hashedVal = this.multHash(key);
+
+    if (!this.buckets[hashedVal]) {
+      let bucketLL = new LinkedList()
+      bucketLL.append({ [key]: value })
+
+      this.buckets[hashedVal] = bucketLL;
+    } else {
+      this.buckets[hashedVal].append({ [key]: value });
+    }
+  }
+
+  get(key) {
+    if (!key) { throw new Error('no key provided for get') }
+    let hashedVal = this.multHash(key);
+    let bucketLL = this.buckets[hashedVal];
+
+    //has a O(1) for append so overall is still O(n) to remove and append order won't matter in hash table
+    let result = HashTable._remove(key, bucketLL);
+    if (!result) {return};
+    bucketLL.append(result);
+    return result[key];
+  }
+
+  remove(key) {
+    if (!key) { throw new Error('no key provided for get') }
+
+    let hashedVal = this.multHash(key);
+    let bucketLL = this.buckets[hashedVal];
+    let result = HashTable._remove(key, bucketLL);
+
+    if(!result){return};
+    return result[key];
+  }
+
+  serialize() {
+    //serialize buckets first
+    this.buckets.forEach(bucket => {
+      if (bucket) {
+        bucket = bucket.serialize();
+      } else { bucket = null }
+    });
+
+    let serializedHashTable = JSON.stringify(this);
+    return serializedHashTable = Buffer.alloc(serializedHashTable.length, serializedHashTable);
+
+  }
+
+  static deserialize(serializedHashTable) {
+    let deserializedHashTable = serializedHashTable.toString();
+    deserializedHashTable = JSON.parse(deserializedHashTable);
+
+    deserializedHashTable.buckets.forEach((bucket, index) => {
+      if (bucket) {
+        let newLinkedList = new LinkedList();
+        newLinkedList.length = bucket.length;
+        newLinkedList.head = bucket.head;
+        newLinkedList.tail = bucket.tail;
+        deserializedHashTable.buckets[index] = newLinkedList;
+      }
+    })
+
+    let newHashTable = new HashTable(deserializedHashTable.nBuckets);
+    newHashTable.buckets = deserializedHashTable.buckets;
+
+    return newHashTable;
+  }
+
+  static _remove(key, bucketLL) {

```

Comment:
Because the logic on this method is a bit confusing/not obvious, it'd be great to provide one-line comments above some of these code blocks explaining what it does.